File: Unique Binary Search Tree.java
public class Solution {
    public int numTrees(int n) {
        if (caches.isEmpty())
        {
            caches.add(new Integer(1));
            caches.add(new Integer(1));
        }
        if (n < caches.size())
            return caches.get(n);
        for (int i = caches.size(); i <= n; i++)
        {
            int num = 0;
            for (int left = 0; left <= i - 1; left++)
            {
                int right = i - 1 - left;
                num += caches.get(left) * caches.get(right);
            }
            caches.add(new Integer(num));
        }
        return caches.get(n);
    }
    static ArrayList<Integer> caches = new ArrayList<Integer>();
}

File: Unique Binary Search Tree.py
class Solution:
	# @return an integer
	def numTrees(self, n):
		if (len(self.caches) == 0):
			self.caches = [1, 1]
		if (n < len(self.caches)):
			return self.caches[n]
		for i in xrange(len(self.caches), n+1):
			num = 0
			for left in xrange(i):
				right = i - 1 - left;
				num += self.caches[left] * self.caches[right]
			self.caches.append(num)
		return self.caches[n]
	caches = []